### PR TITLE
Sockets - Evelynn

### PR DESCRIPTION
**Pull request by evelynnkaplan**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Work.categories returns a list of all categories for media in an array. If I wanted to change the categories available later, I'd only have to change the hardcoded list in this method.
Describe how you approached testing that model method. What edge cases did you come up with? | I had a hard time finding edge cases for this since the data was hardcoded. 
What are `session` and `flash`? What is the difference between them? |  Session is data that the browser stores for the entire time it's open. It helps us keep track of user data. Flash is data stored in a hash that persists through the end of the current request/response cycle and onto the next. It's used to send messages to users. The difference is how long the data is persisted.
What was one thing that you gained more clarity on through this assignment? | Use of more complex and customized routing and customized actions.
What is the Heroku URL of your deployed application? | https://ekaplan-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Rails.application.routes.draw do
+  root "pages#index"
+
+  resources :works
+
+  get "/users/current", to: "users#current_user", as: "current_user"
+
+  resources :users, except: [:new, :create]
```
Comment (droberts-sea) [inline config/routes.rb:8]: This assignment should only need the `index` and `show` actions for users.

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    work.save
+
+    redirect_to works_path
+  end
+
+  def edit
+    if !@work
+      flash[:error] = "No work with that ID found."
+      redirect_to works_path
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:31]: This error checking is repeated many times in this controller. You've already written the `find_work` controller filter, could you do this work there?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class User < ApplicationRecord
+  has_many :votes
+  has_many :works, through: :votes
+
+  def already_voted?(work)
+    return self.works.include?(work)
+  end
```
Comment (droberts-sea) [inline app/models/user.rb:7]: I like that you've made this a model method.

If you wanted to make the interface even simpler / more foolproof, you could define it as a validation on `Vote`, so that it's impossible to create a duplicate vote. Look up "rails validation uniqueness scope"

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    works = Work.all
+    works.max_by { |work| work.votes.count }
+  end
+
+  def self.top_works(category)
+    all_works = Work.where(category: category)
+    unless all_works == []
+      return all_works.max_by(10) { |work| work.votes.count }
+    else
+      return []
```
Comment (droberts-sea) [inline app/models/work.rb:19]: I think that this conditional isn't needed - calling `max_by(10)` on an empty array should return an empty array.

File: app/views/pages/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<body>
+
+  <header>
```
Comment (droberts-sea) [inline app/views/pages/index.html.erb:3]: In general you shouldn't need to include a `<body>` tag in view templates.

File: app/views/pages/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      </p>
+    <% end %>
+  </header>
+
+  <main>
+
+    <% @categories.each do |category|%>
+      <% top_list = Work.top_works(category) %>
+      <% unless top_list == [] %>
+
```
Comment (droberts-sea) [inline app/views/pages/index.html.erb:18]: I'm not a huge fan of calling model methods directly from the view like this. Would it be possible to "tee up" this data in the controller?

File: app/views/pages/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  <main>
+
+    <% @categories.each do |category|%>
+      <% top_list = Work.top_works(category) %>
+      <% unless top_list == [] %>
+        
+        <section>
+          <h3> Top <%= category.capitalize %>s </h3>
+        </section>
```
Comment (droberts-sea) [inline app/views/pages/index.html.erb:21]: You could also call `category.capitalize.pluralize` here.

File: app/views/works/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+     <section class="flash">
+    <% flash.each do |name, message| %>
+      <div class="<%= name %>"><%= message %></div>
+    <% end %>
+  </section>
+
+  <% @categories.each do |category| %>
+    <% all = Work.category_all(category) %>
+    <% unless all == [] %>
+      <h3><%= category.capitalize %></h3>
```
Comment (droberts-sea) [inline app/views/works/index.html.erb:12]: Good work keeping this code DRY (and on the splash page as well)

File: test/controllers/pages_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+require "test_helper"
+
+describe PagesController do
+  describe "index" do
+    it "can get the root path" do
+      get root_path
+      must_respond_with :success
```
Comment (droberts-sea) [inline test/controllers/pages_controller_test.rb:7]: The spotlight presents us with an interesting edge case when there are no works in the database. Do your controller and view handle that without crashing?

File: test/controllers/users_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      get login_path
+
+      must_respond_with :success
+    end
+  end
+
+  describe "login" do
+    it "can login a user" do
+      user = perform_login
+
```
Comment (droberts-sea) [inline test/controllers/users_controller_test.rb:42]: There are a couple of behavior modes you're not covering:
- Logging in a new user (should create that user)
- Attempting to log in with an invalid username (e.g. empty string) (should fail)

File: test/controllers/votes_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      expect(@gump.votes.count).must_equal 1
+      expect(user.votes.count).must_equal 2
+
+      must_respond_with :redirect
+      must_redirect_to work_path(@gump.id)
+    end
+
+    it "won't let a user vote if they're not logged in" do
+      expect { post vote_work_path(@gump.id) }.wont_change "#{@gump.votes.count}"
+      expect(@gump.votes.count).must_equal 0
```
Comment (droberts-sea) [inline test/controllers/votes_controller_test.rb:40]: Good work getting the three interesting cases for this action, especially since the testing logic is so complex.

File: test/models/work_test.rb [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        harry = Work.create!(category: "book", title: title, creator: "JK Rowling", pub_year: 2001, description: "You're a wizard")
+        Vote.create!(work: harry, user: user_one)
+      end
+
+      top_books = Work.top_works("book")
+      expect(top_books.count).must_equal 4
+    end
+
+    it "returns an empty array if there are no works in that category" do
+      top_cars = Work.top_works("car")
```
Comment (droberts-sea) [inline test/models/work_test.rb:90]: Good work finding the edge cases for `top_works`. Another interesting question is, how does it handle ties?

File: test/models/work_test.rb [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    it "can access users who have its votes" do
+      expect(tree.users.count).must_equal 2
+      expect(tree.users.include?(user_one)).must_equal true
+    end
+  end
+
+  describe "categories" do
+    it "must return a list of categories" do
+      categories = Work.categories
+
```
Comment (droberts-sea) [inline test/models/work_test.rb:30]: Does this still work if one of the categories has no works?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions | yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some extra routes
Views are well-organized (DRY, use of semantic HTML, use of partials) | yes
Errors are reported to the user | yes
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | mostly - see inline
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | mostly - see inline
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | no
Look and feel is similar to the original | aside from styling, you've done a good job matching the functionality and workflows of the demo site
**Overall** | Great job overall! Your implementation matches the demo site very closely, and I would say the learning goals for this assignment were definitely met. There are a few places where things could be cleaned up, which I've tried to outline below, but in general I am quite happy with this submission. Keep up the hard work!
